File: Irrgarten_Ruby/Player.rb
#encoding:utf-8

require_relative 'Irrgarten'


module Irrgarten
	class Player
		@@MAX_WEAPONS = 2;
		@@MAX_SHIELDS = 3;
		@@INITIAL_HEALTH = 10;
		@@HITS2LOSE = 3;
		
		def initialize(number, intelligence, strength) #char,float,float
			@number = number
			@intelligence = intelligence
			@strength = strength
			@name = "Player ##{@number}"
			@row = -1
			@col = -1
			@health = @@INITIAL_HEALTH
			@consecutive_hits = 0 #????
			@weapons = []
			@shields = []
		end
		
		attr_reader :row, :col, :number
		
		private
		
		def receive_weapon (w) #weapon
			for i in 0...@weapons.size
				discard = @weapons[i].discard
				
				if (discard)
					@weapons.delete_at(i)
				end
			end
			
			size = @weapons.size
			if (size < @@MAX_WEAPONS)
				@weapons.push(w)
			end
		end
		
		def receive_shield (s) #shield
			for i in 0...@shields.size
				discard = @shields[i].discard
				
				if (discard)
					@shields.delete_at(i)
				end
			end
			
			size = @shields.size
			if (size < @@MAX_SHIELDS)
				@shields.push(s)
			end
		end
		
		def new_weapon ()
			w =  Weapon.new(Dice.weapon_power,Dice.uses_left)
		end
		
		def new_shield ()
			s = Shield.new(Dice.shield_power, Dice.uses_left)
		end
		
		def sum_weapons ()
			suma = 0.0
			for weapon in @weapons
				suma += weapon.attack
			end
			suma	
		end
		
		def sum_shields ()
			suma = 0.0
			for shield in @shields
				suma += shield.attack
			end
			suma
		end
		
		def defensive_energy()
			suma = @intelligence + sum_shields
		end
		
		def manage_hit(received_attack)
			defense = defensive_energy
			lose = false
			
			if (defense < received_attack)
				get_wounded
				inc_consecutive_hits
			else
				reset_hits
			end
			
			if (@consecutive_hits == @@HITS2LOSE || dead)
				reset_hits
				lose=true
			else
				lose=false
			end
			
			lose
		end
		
		def reset_hits ()
			@consecutive_hits = 0
		end
		
		def get_wounded ()
			@health-=1
		end
		
		def inc_consecutive_hits ()
			@consecutive_hits += 1
		end
		
		public
		
		def resurrect() 
			if (dead())
				@health = @@INITIAL_HEALTH
				@weapons.clear()
				@shields.clear()
			end
		end
		
		def set_pos(row, col) #int,int
			@row = row
			@col = col
		end
		
		def dead()
			if (@health <= 0)
				return true
			else
				return false
			end
		end
		
		def move (direction, valid_moves) #Directions, Directions[]
			dir = direction
			
			size = valid_moves.size
			
			contained = valid_moves.include?(direction)
			
			if (!contained && size>0)
				dir = valid_moves[0]
			end
			
			dir
		end	
		
		def attack()
			a = @strength + sum_weapons()
		end
		
		def defend (received_attack) #float
			#Practica 3
		end
		
		def receive_reward()
			w_reward = Dice.weapons_reward
			s_reward = Dice.shields_reward
			
			for i in 0...w_reward
				wnew = new_weapon
				receive_weapon(wnew)
			end
			
			for i in 0...s_reward
				snew = new_shield
				receive_shield(snew)
			end
			
			extra_health = Dice.health_reward
			@health += extra_health
		end
		
		def to_s()
			"Name: #{@name}, Number: #{@number}, Intelligence: #{@intelligence}, Strength: #{@strength}, Health: #{@health}, Position: (#{@row},#{@col}), ConsecutiveHits: #{@consecutive_hits} \n Armas: \n #{@weapons} \n Escudos: \n #{@shields}"
		end
	end
end

File: Irrgarten/src/irrgarten/Player.java

package irrgarten;

import java.util.ArrayList;

public class Player {
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    private String name;
    private char number;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    private int consecutiveHits=0;
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    
    
    private void receiveWeapon( Weapon w){ 
        for(int i=0; i< weapons.size(); i++){
            boolean discard=weapons.get(i).discard();
            
            if (discard){
                weapons.remove(i);
            }
        }
        
        int size= weapons.size();
        if (size< MAX_WEAPONS){ 
            weapons.add(w);
        }
    }
    
    private void receiveShield(Shield s){ 
        for(int i=0; i< shields.size(); i++){
            boolean discard=shields.get(i).discard();
            
            if (discard){
                shields.remove(i);
            }
        }
        
        int size= shields.size();
        if (size< MAX_SHIELDS){ 
            shields.add(s);
        }
    }
    
    private Weapon newWeapon(){
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }
    
    private Shield newShield(){
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }
    
    private float sumWeapons(){
        float suma = 0.0f;
        for (int i=0; i< weapons.size(); i++){
            suma += weapons.get(i).attack();
        }
        return suma;
    }
    
    private float sumShields(){
        float suma = 0.0f;
        for (int i=0; i< shields.size(); i++){
            suma += shields.get(i).protect();
        }
        return suma;
    }
    
    private float defensiveEnergy(){
        return intelligence + sumShields();
    }
    
    private boolean manageHit(float receivedAttack){
        float defense= defensiveEnergy();
        boolean lose=false;
        
        if(defense<receivedAttack){
            getWounded();
            incConsecutiveHits();
        } else {
            resetHits();
        }

        if(consecutiveHits==HITS2LOSE || (dead())) {
            resetHits();
            lose=true;
        } else {
            lose=false;
        }
        return lose;
    }
    
    private void resetHits(){
        consecutiveHits=0;
    }
    
    private void getWounded(){
        health--;
    }
    
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
    
    public Player ( char number , float intelligence, float strength){
           this.number=number;
           this.intelligence=intelligence;
           this.strength=strength;
           this.name = "Player #"+number;
           this.row=-1;
           this.col=-1;
           this.health=INITIAL_HEALTH;
           this.weapons = new ArrayList<>();
           this.shields = new ArrayList<>();
    }
    
    public void resurrect(){
        assert dead() : "Estas intentando resucitar sin estar muerto";
       
        this.health=INITIAL_HEALTH;
        this.weapons.clear();
        this.shields.clear();
    }
    
    public int getRow(){
        return this.row;
    }
    
    public int getCol(){
        return this.col;
    }  
    
    public char getNumber(){
        return this.number;
    }
    
    public void setPos(int row, int col){
        this.row=row;
        this.col= col;
    }
    
    public boolean dead(){
        if (this.health <=0) {
                return true;
        } else {
                return false;
        }
    }
    
    public Directions move ( Directions direction, ArrayList<Directions> validMoves){ 
        Directions dir=direction;
        
        int size= validMoves.size();
       
        boolean contained=validMoves.contains(direction); 
        
        if (!contained && size>0){
            dir=validMoves.get(0);
        }
        
          return dir;
    }
    
    public float attack(){
        return this.strength + sumWeapons();
    }
    
    public boolean defend(float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        int wReward= Dice.weaponsReward();
        int sReward= Dice.shieldsReward();
        
        for (int i=0; i< wReward; i++){
            Weapon wnew= newWeapon();
            this.receiveWeapon(wnew);
        }
        for (int i=0; i< sReward; i++){
            Shield snew= newShield();
            this.receiveShield(snew);
        }
        int extraHealth= Dice.healthReward();
        this.health += extraHealth;
    }
    
    @Override
    public String toString(){
        return ("Name: " + name + ", Number: " + number + ", Intelligence: " + intelligence + ", Strength: " + strength + 
                ", Health: " + health + ", Position: (" + row + "," + col + ") , ConsecutiveHits: " + consecutiveHits + " \nArmas: \n" + weapons + "\nEscudos: \n" + shields);
    }
    
    
    
}
